File: day11/Puzzle.cs
using System.Collections.Generic;

namespace Aoc2021
{
    public class Puzzle : AocPuzzle<int[], int>
    {
        protected override IDictionary<string, int> TestAnswers { get; } = new Dictionary<string, int>
        {
            { "part1", 1656 },
            { "part2", 195 }
        };

        protected override int[] ParseInput(string[] lines)
        {
            var values = new int[SquidGrid.Size * SquidGrid.Size];
            int offset = 0;
            int zero = (int)'0';
            foreach (string line in lines)
            {
                for (int i = 0; i < SquidGrid.Size; ++i)
                {
                    values[offset + i] = line[i] - zero;
                }
                offset += SquidGrid.Size;
            }

            return values;
        }

        protected override int RunPart1()
        {
            var squidGrid = new SquidGrid(_input);
            for (int i = 0; i < 100; ++i)
            {
                squidGrid.ModelStep();
            }

            return squidGrid.FlashCount;
        }

        protected override int RunPart2()
        {
            var squidGrid = new SquidGrid(_input);
            int step = 1;
            while (squidGrid.ModelStep() < _input.Length)
            {
                ++step;
            }

            return step;
        }
    }
}

File: day11-js/puzzle.js
import { createAocPuzzle } from "./aoc.js";
import { createSquidGrid, SIZE } from "./squid-grid.js";

const testAnswers = {
  part1: 1656,
  part2: 195
};

function parseInput(lines) {
  const values = new Int8Array(SIZE * SIZE);
  let offset = 0;
  for (let y = 0; y < SIZE; ++y) {
    for (let x = 0; x < SIZE; ++x) {
      values[offset + x] = +lines[y][x];
    }
    offset += SIZE;
  }
  return values;
}

function runPart1(input) {
  const squidGrid = createSquidGrid(input);
  for (let i = 0; i < 100; ++i) {
    squidGrid.modelStep();
  }
  return squidGrid.flashCount;
}

function runPart2(input) {
  const squidGrid = createSquidGrid(input);
  let step = 1;
  while (squidGrid.modelStep() < input.length) {
    ++step;
  }
  return step;
}

export function createPuzzle() {
  return createAocPuzzle({ parseInput, runPart1, runPart2, testAnswers });
}
